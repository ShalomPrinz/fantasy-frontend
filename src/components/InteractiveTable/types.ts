interface TableColumn {
  id: number;
  label: string;
  path?: string;
  content?: Function;
}

export { TableColumn };
