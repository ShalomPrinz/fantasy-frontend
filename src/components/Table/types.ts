interface TableColumn {
  id: number;
  path?: string;
  content?: Function;
}

export { TableColumn };
