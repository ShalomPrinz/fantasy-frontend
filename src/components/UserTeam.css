.bg-team-form {
    background: linear-gradient(#b6f4fc, #9198e5, #e66465);
}