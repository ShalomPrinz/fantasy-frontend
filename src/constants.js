export const FIELD_IMAGE_DEFAULT_WIDTH = 550;
export const FIELD_LAYOUT_MIN_WIDTH = 800;
