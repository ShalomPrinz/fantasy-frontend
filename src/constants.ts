export const FIELD_IMAGE_DEFAULT_WIDTH = 550;
export const FIELD_LAYOUT_MIN_WIDTH = 1000;
export const FULL_SCREEN_MIN_WIDTH = 1200;

// My Backend Endpoints
export const baseURL = "http://localhost:8080";
export const playersUrl = "/players/query";
export const usernamesUrl = "/users/query";
export const registerUrl = "/register";
export const userInfoUrl = "/user";
export const addPlayerUrl = "/user/addplayer";
export const removePlayerUrl = "/user/removeplayer";
export const acceptLeagueInviteUrl = "/accept/leagueinvite";
export const rejectLeagueInviteUrl = "/reject/leagueinvite";
