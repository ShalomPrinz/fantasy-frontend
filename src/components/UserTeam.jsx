import React from "react";

import TeamLayout from "./TeamLayout";
import "./UserTeam.css";

const UserTeam = ({ user }) => {

  if (!user) {
      return <h1 className="text-center m-5 p-5 bg-danger text-white">Please Log In to view your team</h1>
  }

  if (!user.team) {
      user.team = {
          GK: [{name: "Ter Stegen", team: "Barcelona"}],
          DEF: [{name: "Pique", team: "Barcelona"}, {name: "Kounde", team: "Barcelona"}, {name: "Araujo", team: "Barcelona"}],
          MID: [{name: "Bellingham", team: "Dortmund"}],
          ATT: [{name: "Mokuku", team: "Dortmund"}]
      }
  }
  
  return (
    <>
        <main className="container w-75 p-4 mt-4 rounded bg-team-form">
            <h2 className="fw-bold"> {user.name}'s Team </h2>
            Select a maximum of 3 players of a single team
            <div className="bg-white text-center shadow-below rounded my-3">
                <h5 className="bg-default py-2 mx-3 rounded-bottom"> Gameweek 1 </h5>
                <div className="centered-flex">
                    Gameweek 1 deadline: 
                    <div className="fw-bold ps-2"> Friday 20:30 </div>
                </div>
                <h2 className="py-3 centered-flex">
                    Players Selected:
                    <div className="bg-light-danger mx-3 p-2 rounded"> 0 / 11 </div>
                </h2>
            </div>
            <TeamLayout team={user.team} />
        </main>
    </>
  );
};

export default UserTeam;