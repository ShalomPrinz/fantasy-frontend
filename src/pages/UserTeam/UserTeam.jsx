import React, { useState } from "react";

import { faCirclePlus } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";
import { toast } from "react-toastify";

import { PlayerJersey, Search, TabChoice, Table, TeamLayout, TeamList } from "../../components";
import { FIELD_LAYOUT_MIN_WIDTH } from '../../constants'; 
import useWindowSize from "../../hooks/useWindowSize";
import "./UserTeam.css";

const UserTeam = ({ user }) => {
  const { width } = useWindowSize()
  const [query, setQuery] = useState("");

  if (!user) {
      return <h1 className="text-center m-5 p-5 bg-danger text-white">Please Log In to view your team</h1>
  }

  if (!user.team) {
      user.team = [
          {
              id: 0,
              label: 'GK',
              players: [{id: 0, name: "Ter Stegen", team: "Barcelona"}]
          },
          {
              id: 1,
              label: 'DEF',
              players: [{id: 1, name: "Pique", team: "Barcelona"}, {id: 2, name: "Kounde", team: "Barcelona"}, {id: 3, name: "Araujo", team: "Barcelona"}]
          },
          {
              id: 2,
              label: 'MID',
              players: [{id: 4, name: "Bellingham", team: "Dortmund"}]
          },
          {
              id: 3,
              label: 'ATT',
              players: [{id: 5, name: "Mokuku", team: "Dortmund"}]
          }
      ]
  }

  const chosenPlayers = user.team.reduce((sum, { players }) => sum + players.length, 0)
  const selectedPlayersBackground = () => chosenPlayers === 11 ? "success" : "danger"

  const tabs = [
        {
            id: 0,
            label: 'Field',
            Component: <TeamLayout team={user.team}/>,
            OnClick: () => {
                if (width <= FIELD_LAYOUT_MIN_WIDTH)
                    toast.warn('Your screen is too small')
            }
        },
        {
            id: 1,
            label: 'List',
            Component: <TeamList team={user.team} />
        }
  ]

  const columns = [
    {
        id: 0,
        path: "team",
        content: (p) => <PlayerJersey team={p.team} width="60px" />
    },
    {
        id: 1,
        path: "name",
        content: (p) => (
            <>
                <div className="fs-4">{p.name}</div>
                <div className="text-muted">{p.team}, {p.role}</div>
            </>
        )
    },
    {
        id: 2,
        content: () => (
            <div className="text-center">
                <FontAwesomeIcon className="fa-3x clickable text-primary" icon={faCirclePlus} />
            </div>
        )
    }
  ]

  const data = [
    {
        id: 0,
        props: {
            name: "Pena",
            role: "GK",
            team: "Barcelona"
        }
    },
    {
        id: 1,
        props: {
            name: "Hummels",
            role: "DEF",
            team: "Dortmund"
        }
    },
    {
        id: 2,
        props: {
            name: "Pedri",
            role: "MID",
            team: "Barcelona"
        }
    },
    {
        id: 3,
        props: {
            name: "Dembele",
            role: "ATT",
            team: "Barcelona"
        }
    },
    {
        id: 4,
        props: {
            name: "Ansu Fati",
            role: "ATT",
            team: "Barcelona"
        }
    },
    {
        id: 5,
        props: {
            name: "Mokuku",
            role: "ATT",
            team: "Dortmund"
        }
    }
  ]
  
  return (
    <main className="container p-4 my-4 rounded bg-team-form">
        <Row>
            <Col>
                <h2 className="fw-bold"> {user.name}'s Team </h2>
                Select a maximum of 3 players of a single team
            </Col>
        </Row>
        <Row>
            <Col className={`${width < FIELD_LAYOUT_MIN_WIDTH ? 'w-100' : 'pe-5'}`} sm="8">

                <div className="bg-white text-center shadow-below rounded my-3">
                    <h5 className="bg-default py-2 mx-3 rounded-bottom"> Gameweek 1 </h5>
                    <div className="centered-flex">
                        Gameweek 1 deadline: 
                        <div className="fw-bold ps-2"> Friday 20:30 </div>
                    </div>
                    <h2 className="py-3 centered-flex">
                        Players Selected:
                        <div className={`bg-light-${selectedPlayersBackground()} mx-3 p-2 rounded`}> {chosenPlayers} / 11 </div>
                    </h2>
                </div>

                <TabChoice tabs={tabs} defaultTab={tabs[0]} />
            </Col>
            <Col className="ps-1" >
                <Search onChange={(v) => setQuery(v)} value={query} />
                <Table className="bg-white" columns={columns} data={data} />
            </Col>
        </Row>
    </main>
  );
};

export default UserTeam;