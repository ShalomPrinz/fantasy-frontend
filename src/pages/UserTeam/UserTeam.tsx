import Col from "react-bootstrap/Col";
import Row from "react-bootstrap/Row";

import { PlayersTable, TeamTabChoice } from "../../components";
import { FULL_SCREEN_MIN_WIDTH } from "../../constants";
import { useUser } from "../../contexts/UserContext";
import TeamProvider, { useTeamState } from "../../contexts/UserTeamContext";
import useWindowWidth from "../../hooks/useWindowWidth";
import "./UserTeam.css";

interface UserTeamProps {
  name: string;
}

const UserTeam = ({ name }: UserTeamProps) => {
  const { count: playersCount } = useTeamState();

  const width = useWindowWidth();
  const contentStyle = width < FULL_SCREEN_MIN_WIDTH ? "mx-auto" : "pe-5";

  const selectedPlayersBackground = playersCount === 11 ? "success" : "danger";

  return (
    <main className="container p-4 my-4 rounded bg-team-form">
      <Row>
        <Col>
          <h2 className="fw-bold"> {name}'s Team </h2>
          Select a maximum of 3 players of a single team
        </Col>
      </Row>
      <Row>
        <Col className={contentStyle} sm="8">
          <div className="bg-white text-center shadow-below rounded my-3">
            <h5 className="bg-default py-2 mx-3 rounded-bottom">Gameweek 1</h5>
            <div className="centered-flex">
              Gameweek 1 deadline:
              <div className="fw-bold ps-2"> Friday 20:30 </div>
            </div>
            <h2 className="py-3 centered-flex">
              Players Selected:
              <div
                className={`bg-light-${selectedPlayersBackground} mx-3 p-2 rounded`}
              >
                {playersCount} / 11
              </div>
            </h2>
          </div>

          <TeamTabChoice />
        </Col>
        <Col className="ps-1">
          <PlayersTable />
        </Col>
      </Row>
    </main>
  );
};

const UserTeamWrapper = () => {
  const { loading, user } = useUser();

  const Message = (color: string, text: string) => (
    <h1 className={`text-center my-5 p-5 bg-${color} text-white w-75 mx-auto`}>
      {text}
    </h1>
  );

  if (typeof user === "undefined") {
    return loading
      ? Message("info", "Loading...")
      : Message("danger", "Please Log In to view your team");
  } else
    return (
      <TeamProvider>
        <UserTeam name={user.name} />
      </TeamProvider>
    );
};

export default UserTeamWrapper;
