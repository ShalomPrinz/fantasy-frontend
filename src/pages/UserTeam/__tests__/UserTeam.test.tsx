import {
  clickElement,
  renderQueryClient,
  screen,
  setWindowSize,
  sleep,
} from "setupTests";

import { FIELD_LAYOUT_MIN_WIDTH } from "../../../constants";
import { Player, Team, User } from "../../../types";
import UserTeam from "../UserTeam";

describe("UserTeam", () => {
  it("should render UserTeam component", async () => {
    const players: Player[] = [
      { id: 0, name: "Messi", role: "ATT", team: "Barcelona" },
    ];
    const user = new User("Some Name", new Team(players));

    const { asFragment } = renderQueryClient(<UserTeam user={user} />);
    await sleep();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should not allow field layout if window size is smaller than minimum", async () => {
    const players: Player[] = [
      { id: 0, name: "Messi", role: "ATT", team: "Barcelona" },
    ];
    const appUser = new User("Some Name", new Team(players));

    setWindowSize(FIELD_LAYOUT_MIN_WIDTH - 1);
    const { asFragment, user } = renderQueryClient(<UserTeam user={appUser} />);
    await sleep();
    const firstRender = asFragment();

    const element = screen.getByRole("button", { name: /Field/i });
    await clickElement(user, element);

    expect(asFragment()).toEqual(firstRender);
  });
});
