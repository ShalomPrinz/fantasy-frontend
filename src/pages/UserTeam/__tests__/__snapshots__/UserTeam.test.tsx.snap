// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTeam should render UserTeam component 1`] = `
<DocumentFragment>
  <main
    class="container p-4 my-4 rounded bg-team-form"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <h2
          class="fw-bold"
        >
           Some Name's Team 
        </h2>
        Select a maximum of 3 players of a single team
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="pe-5 col-sm-8"
      >
        <div
          class="bg-white text-center shadow-below rounded my-3"
        >
          <h5
            class="bg-default py-2 mx-3 rounded-bottom"
          >
            Gameweek 1
          </h5>
          <div
            class="centered-flex"
          >
            Gameweek 1 deadline:
            <div
              class="fw-bold ps-2"
            >
               Friday 20:30 
            </div>
          </div>
          <h2
            class="py-3 centered-flex"
          >
            Players Selected:
            <div
              class="bg-light-danger mx-3 p-2 rounded"
            >
              1 / 11
            </div>
          </h2>
        </div>
        <div
          class="centered-flex py-4"
        >
          <button
            class="fs-2 bg-default py-2 px-5 tab-choice"
            type="button"
          >
            Field
          </button>
          <button
            class="fs-2 bg-default py-2 px-5 tab-choice"
            type="button"
          >
            List
          </button>
        </div>
        <div
          class="position-relative"
        >
          <img
            alt="Team Field Background"
            class="w-100 rounded"
            src="field.png"
          />
          <div
            class="position-absolute p-5 top-0 centered-flex flex-column h-100 overflow-hidden container"
          >
            <div
              class="my-1 row"
            />
            <div
              class="my-1 row"
            />
            <div
              class="my-1 row"
            />
            <div
              class="my-1 row"
            >
              <div
                class="mx-auto col"
              >
                <div
                  class="text-center"
                  style="width: 77px;"
                >
                  <img
                    class=""
                    src="Barcelona.png"
                    style="width: 55px;"
                    title="Barcelona"
                  />
                  <div
                    class="text-truncate bg-default rounded p-1 w-100"
                  >
                    Messi
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ps-1 col"
      >
        <input
          class="form-control my-3 fs-1 p-4"
          placeholder="Search..."
          type="text"
          value=""
        />
        <div
          class="table-responsive mt-0"
        >
          <table
            class="table rounded bg-white"
          >
            <tbody>
              <tr>
                <td
                  class="align-middle"
                >
                  <img
                    class=""
                    src="Barcelona.png"
                    style="width: 60px;"
                    title="Barcelona"
                  />
                </td>
                <td
                  class="align-middle"
                >
                  <div
                    class="fs-4"
                  >
                    Ter Stegen
                  </div>
                  <div
                    class="text-muted"
                  >
                    Barcelona, GK
                  </div>
                </td>
                <td
                  class="align-middle"
                >
                  <div
                    class="text-center"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-circle-plus fa-3x clickable text-primary"
                      data-icon="circle-plus"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 512c141.4 0 256-114.6 256-256S397.4 0 256 0S0 114.6 0 256S114.6 512 256 512zM232 344V280H168c-13.3 0-24-10.7-24-24s10.7-24 24-24h64V168c0-13.3 10.7-24 24-24s24 10.7 24 24v64h64c13.3 0 24 10.7 24 24s-10.7 24-24 24H280v64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
